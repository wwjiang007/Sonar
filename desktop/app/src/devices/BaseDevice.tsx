/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import stream from 'stream';
import {FlipperDevicePlugin} from 'flipper';
import {sortPluginsByName} from '../utils/pluginUtils';

export type LogLevel =
  | 'unknown'
  | 'verbose'
  | 'debug'
  | 'info'
  | 'warn'
  | 'error'
  | 'fatal';

export type DeviceLogEntry = {
  readonly date: Date;
  readonly pid: number;
  readonly tid: number;
  readonly app?: string;
  readonly type: LogLevel;
  readonly tag: string;
  readonly message: string;
};

export type DeviceShell = {
  stdout: stream.Readable;
  stderr: stream.Readable;
  stdin: stream.Writable;
};

export type DeviceLogListener = (entry: DeviceLogEntry) => void;

export type DeviceType =
  | 'emulator'
  | 'physical'
  | 'archivedEmulator'
  | 'archivedPhysical';

export type DeviceExport = {
  os: OS;
  title: string;
  deviceType: DeviceType;
  serial: string;
  logs: Array<DeviceLogEntry>;
};

export type OS = 'iOS' | 'Android' | 'Windows' | 'MacOS' | 'JSWebApp' | 'Metro';

export default class BaseDevice {
  constructor(serial: string, deviceType: DeviceType, title: string, os: OS) {
    this.serial = serial;
    this.title = title;
    this.deviceType = deviceType;
    this.os = os;
  }

  // operating system of this device
  os: OS;

  // human readable name for this device
  title: string;

  // type of this device
  deviceType: DeviceType;

  // serial number for this device
  serial: string;

  // possible src of icon to display next to the device title
  icon: string | null | undefined;

  logListeners: Map<Symbol, DeviceLogListener> = new Map();
  logEntries: Array<DeviceLogEntry> = [];
  isArchived: boolean = false;
  // if imported, stores the original source location
  source = '';

  // sorted list of supported device plugins
  devicePlugins!: string[];

  supportsOS(os: OS) {
    return os.toLowerCase() === this.os.toLowerCase();
  }

  displayTitle(): string {
    return this.title;
  }

  toJSON(): DeviceExport {
    return {
      os: this.os,
      title: this.title,
      deviceType: this.deviceType,
      serial: this.serial,
      logs: this.getLogs(),
    };
  }

  teardown() {}

  supportedColumns(): Array<string> {
    return ['date', 'pid', 'tid', 'tag', 'message', 'type', 'time'];
  }

  addLogListener(callback: DeviceLogListener): Symbol {
    const id = Symbol();
    this.logListeners.set(id, callback);
    return id;
  }

  _notifyLogListeners(entry: DeviceLogEntry) {
    if (this.logListeners.size > 0) {
      this.logListeners.forEach((listener) => {
        // prevent breaking other listeners, if one listener doesn't work.
        try {
          listener(entry);
        } catch (e) {
          console.error(`Log listener exception:`, e);
        }
      });
    }
  }

  addLogEntry(entry: DeviceLogEntry) {
    this.logEntries.push(entry);
    this._notifyLogListeners(entry);
  }

  getLogs(startDate: Date | null = null) {
    return startDate != null
      ? this.logEntries.filter((log) => {
          return log.date > startDate;
        })
      : this.logEntries;
  }

  clearLogs(): Promise<void> {
    // Only for device types that allow clearing.
    this.logEntries = [];
    return Promise.resolve();
  }

  removeLogListener(id: Symbol) {
    this.logListeners.delete(id);
  }

  navigateToLocation(_location: string) {
    throw new Error('unimplemented');
  }

  archive(): any | null | undefined {
    return null;
  }

  screenshot(): Promise<Buffer> {
    return Promise.reject(
      new Error('No screenshot support for current device'),
    );
  }

  async screenCaptureAvailable(): Promise<boolean> {
    return false;
  }

  async startScreenCapture(_destination: string): Promise<void> {
    throw new Error('startScreenCapture not implemented on BaseDevice ');
  }

  async stopScreenCapture(): Promise<string | null> {
    return null;
  }

  loadDevicePlugins(devicePlugins?: Map<string, typeof FlipperDevicePlugin>) {
    this.devicePlugins = Array.from(devicePlugins ? devicePlugins.values() : [])
      .filter((plugin) => plugin.supportsDevice(this))
      .sort(sortPluginsByName)
      .map((plugin) => plugin.id);
  }
}
