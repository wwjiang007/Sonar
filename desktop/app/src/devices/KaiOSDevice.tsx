/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import AndroidDevice from './AndroidDevice';

export default class KaiOSDevice extends AndroidDevice {
  async screenCaptureAvailable() {
    // The default way of capturing screenshots through adb does not seem to work
    // There is a way of getting a screenshot through KaiOS dev tools though
    return false;
  }
}
