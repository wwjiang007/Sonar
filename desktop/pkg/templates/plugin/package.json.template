{
  "$schema": "https://fbflipper.com/schemas/plugin-package/v2.json",
  "name": "{{package_name}}",
  "id": "{{id}}",
  "version": "1.0.0",
  "main": "dist/bundle.js",
  "flipperBundlerEntry": "src/index.tsx",
  "license": "MIT",
  "keywords": [
    "flipper-plugin"
  ],
  "icon": "apps",
  "title": "{{title}}",
  "scripts": {
    "lint": "flipper-pkg lint",
    "prepack": "flipper-pkg lint && flipper-pkg bundle",
    "build": "flipper-pkg bundle",
    "watch": "flipper-pkg bundle --watch"
  },
  "peerDependencies": {
    "flipper": "latest"
  },
  "devDependencies": {
    "@types/react": "latest",
    "@types/react-dom": "latest",
    "flipper": "latest",
    "flipper-pkg": "latest"
  }
}
