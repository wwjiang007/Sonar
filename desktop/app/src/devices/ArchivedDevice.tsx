/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import BaseDevice from './BaseDevice';
import {DeviceType, OS, DeviceShell, DeviceLogEntry} from './BaseDevice';
import {SupportFormRequestDetailsState} from '../reducers/supportForm';

function normalizeArchivedDeviceType(deviceType: DeviceType): DeviceType {
  let archivedDeviceType = deviceType;
  if (archivedDeviceType === 'emulator') {
    archivedDeviceType = 'archivedEmulator';
  } else if (archivedDeviceType === 'physical') {
    archivedDeviceType = 'archivedPhysical';
  }
  return archivedDeviceType;
}

export default class ArchivedDevice extends BaseDevice {
  constructor(options: {
    serial: string;
    deviceType: DeviceType;
    title: string;
    os: OS;
    logEntries: Array<DeviceLogEntry>;
    screenshotHandle: string | null;
    source?: string;
    supportRequestDetails?: SupportFormRequestDetailsState;
  }) {
    super(
      options.serial,
      normalizeArchivedDeviceType(options.deviceType),
      options.title,
      options.os,
    );
    this.logs = options.logEntries;
    this.source = options.source || '';
    this.supportRequestDetails = options.supportRequestDetails;
    this.archivedScreenshotHandle = options.screenshotHandle;
  }

  logs: Array<DeviceLogEntry>;
  archivedScreenshotHandle: string | null;
  isArchived = true;

  displayTitle(): string {
    return `${this.title} ${this.source ? '(Imported)' : '(Offline)'}`;
  }

  supportRequestDetails?: SupportFormRequestDetailsState;

  getLogs() {
    return this.logs;
  }

  clearLogs(): Promise<void> {
    this.logs = [];
    return Promise.resolve();
  }

  spawnShell(): DeviceShell | undefined | null {
    return null;
  }

  getArchivedScreenshotHandle(): string | null {
    return this.archivedScreenshotHandle;
  }
}
