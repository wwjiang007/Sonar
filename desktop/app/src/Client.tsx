/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {FlipperPlugin, FlipperDevicePlugin} from './plugin';
import BaseDevice, {OS} from './devices/BaseDevice';
import {App} from './App';
import {Logger} from './fb-interfaces/Logger';
import {Store} from './reducers/index';
import {setPluginState} from './reducers/pluginStates';
import {Payload, ConnectionStatus} from 'rsocket-types';
import {Flowable, Single} from 'rsocket-flowable';
import {performance} from 'perf_hooks';
import {reportPlatformFailures, reportPluginFailures} from './utils/metrics';
import {notNull} from './utils/typeUtils';
import {default as isProduction} from './utils/isProduction';
import {registerPlugins} from './reducers/plugins';
import createTableNativePlugin from './plugins/TableNativePlugin';
import {EventEmitter} from 'events';
import invariant from 'invariant';
import {flipperRecorderAddEvent} from './utils/pluginStateRecorder';
import {
  getPluginKey,
  defaultEnabledBackgroundPlugins,
} from './utils/pluginUtils';
import {processMessageLater} from './utils/messageQueue';
import {sideEffect} from './utils/sideEffect';
import {emitBytesReceived} from './dispatcher/tracking';

type Plugins = Array<string>;

export type ClientQuery = {
  app: string;
  os: OS;
  device: string;
  device_id: string;
  sdk_version?: number;
};

export type ClientExport = {
  id: string;
  query: ClientQuery;
};

type ErrorType = {message: string; stacktrace: string; name: string};
type Params = {
  api: string;
  method: string;
  params?: Object;
};
type RequestMetadata = {method: string; id: number; params: Params | undefined};

const handleError = (store: Store, device: BaseDevice, error: ErrorType) => {
  if (isProduction()) {
    return;
  }
  const crashReporterPlugin = store
    .getState()
    .plugins.devicePlugins.get('CrashReporter');
  if (!crashReporterPlugin) {
    return;
  }

  const pluginKey = getPluginKey(null, device, 'CrashReporter');

  const persistedState = {
    ...crashReporterPlugin.defaultPersistedState,
    ...store.getState().pluginStates[pluginKey],
  };
  const isCrashReport: boolean = Boolean(error.name || error.message);
  const payload = isCrashReport
    ? {
        name: error.name,
        reason: error.message,
        callstack: error.stacktrace,
      }
    : {
        name: 'Plugin Error',
        reason: JSON.stringify(error),
      };

  const newPluginState =
    crashReporterPlugin.persistedStateReducer == null
      ? persistedState
      : crashReporterPlugin.persistedStateReducer(
          persistedState,
          'flipper-crash-report',
          payload,
        );
  if (persistedState !== newPluginState) {
    store.dispatch(
      setPluginState({
        pluginKey,
        state: newPluginState,
      }),
    );
  }
};

export interface FlipperClientConnection<D, M> {
  connectionStatus(): Flowable<ConnectionStatus>;
  close(): void;
  fireAndForget(payload: Payload<D, M>): void;
  requestResponse(payload: Payload<D, M>): Single<Payload<D, M>>;
}

export default class Client extends EventEmitter {
  app: App | undefined;
  connected: boolean;
  id: string;
  query: ClientQuery;
  sdkVersion: number;
  messageIdCounter: number;
  plugins: Plugins;
  backgroundPlugins: Plugins;
  connection: FlipperClientConnection<any, any> | null | undefined;
  store: Store;
  activePlugins: Set<string>;
  device: Promise<BaseDevice>;
  _deviceResolve: (device: BaseDevice) => void = (_) => {};
  logger: Logger;
  lastSeenDeviceList: Array<BaseDevice>;
  broadcastCallbacks: Map<string, Map<string, Set<Function>>>;

  requestCallbacks: Map<
    number,
    {
      resolve: (data: any) => void;
      reject: (err: Error) => void;
      metadata: RequestMetadata;
      // eslint-disable-next-line prettier/prettier
    }
  >;

  constructor(
    id: string,
    query: ClientQuery,
    conn: FlipperClientConnection<any, any> | null | undefined,
    logger: Logger,
    store: Store,
    plugins?: Plugins | null | undefined,
    device?: BaseDevice,
  ) {
    super();
    this.connected = true;
    this.plugins = plugins ? plugins : [];
    this.backgroundPlugins = [];
    this.connection = conn;
    this.id = id;
    this.query = query;
    this.sdkVersion = query.sdk_version || 0;
    this.messageIdCounter = 0;
    this.logger = logger;
    this.store = store;
    this.broadcastCallbacks = new Map();
    this.requestCallbacks = new Map();
    this.activePlugins = new Set();
    this.lastSeenDeviceList = [];

    this.device = device
      ? Promise.resolve(device)
      : new Promise((resolve, _reject) => {
          this._deviceResolve = resolve;
        });

    const client = this;
    if (conn) {
      conn.connectionStatus().subscribe({
        onNext(payload) {
          if (payload.kind == 'ERROR' || payload.kind == 'CLOSED') {
            client.connected = false;
          }
        },
        onSubscribe(subscription) {
          subscription.request(Number.MAX_SAFE_INTEGER);
        },
        onError(payload) {
          console.error('[client] connection status error ', payload);
        },
      });
    }
  }

  /* All clients should have a corresponding Device in the store.
     However, clients can connect before a device is registered, so wait a
     while for the device to be registered if it isn't already. */
  async setMatchingDevice(): Promise<void> {
    return reportPlatformFailures(
      new Promise<BaseDevice>((resolve, reject) => {
        let unsubscribe: () => void = () => {};

        const device = this.store
          .getState()
          .connections.devices.find(
            (device) => device.serial === this.query.device_id,
          );
        if (device) {
          resolve(device);
          return;
        }

        const timeout = setTimeout(() => {
          unsubscribe();
          const error = `Timed out waiting for device for client ${this.id}`;
          console.error(error);
          reject(error);
        }, 5000);
        unsubscribe = sideEffect(
          this.store,
          {name: 'waitForDevice', throttleMs: 100},
          (state) => state.connections.devices,
          (newDeviceList) => {
            if (newDeviceList === this.lastSeenDeviceList) {
              return;
            }
            this.lastSeenDeviceList = newDeviceList;
            const matchingDevice = newDeviceList.find(
              (device) => device.serial === this.query.device_id,
            );
            if (matchingDevice) {
              clearTimeout(timeout);
              resolve(matchingDevice);
              unsubscribe();
            }
          },
        );
      }),
      'client-setMatchingDevice',
    ).then((device) => {
      this._deviceResolve(device);
    });
  }

  supportsPlugin(Plugin: typeof FlipperPlugin): boolean {
    return this.plugins.includes(Plugin.id);
  }

  isBackgroundPlugin(pluginId: string) {
    return this.backgroundPlugins.includes(pluginId);
  }

  shouldConnectAsBackgroundPlugin(pluginId: string) {
    return (
      defaultEnabledBackgroundPlugins.includes(pluginId) ||
      this.store
        .getState()
        .connections.userStarredPlugins[this.query.app]?.includes(pluginId)
    );
  }

  async init() {
    this.setMatchingDevice();
    await this.loadPlugins();
    this.backgroundPlugins = await this.getBackgroundPlugins();
    this.backgroundPlugins.forEach((plugin) => {
      if (this.shouldConnectAsBackgroundPlugin(plugin)) {
        this.initPlugin(plugin);
      }
    });
  }

  // get the supported plugins
  async loadPlugins(): Promise<Plugins> {
    const plugins = await this.rawCall<{plugins: Plugins}>(
      'getPlugins',
      false,
    ).then((data) => data.plugins);
    this.plugins = plugins;
    const nativeplugins = plugins
      .map((plugin) => /_nativeplugin_([^_]+)_([^_]+)/.exec(plugin))
      .filter(notNull)
      .map(([id, type, title]) => {
        // TODO put this in another component, and make the "types" registerable
        switch (type) {
          case 'Table':
            return createTableNativePlugin(id, title);
          default: {
            return null;
          }
        }
      })
      .filter(Boolean);
    this.store.dispatch(registerPlugins(nativeplugins as any));
    return plugins;
  }

  // get the supported background plugins
  async getBackgroundPlugins(): Promise<Plugins> {
    if (this.sdkVersion < 4) {
      return [];
    }
    return await this.rawCall<{plugins: Plugins}>(
      'getBackgroundPlugins',
      false,
    ).then((data) => data.plugins);
  }

  // get the plugins, and update the UI
  async refreshPlugins() {
    const oldBackgroundPlugins = this.backgroundPlugins;
    await this.loadPlugins();
    const newBackgroundPlugins = await this.getBackgroundPlugins();
    this.backgroundPlugins = newBackgroundPlugins;
    // diff the background plugin list, disconnect old, connect new ones
    oldBackgroundPlugins.forEach((plugin) => {
      if (
        !newBackgroundPlugins.includes(plugin) &&
        this.store
          .getState()
          .connections.userStarredPlugins[this.query.app]?.includes(plugin)
      ) {
        this.deinitPlugin(plugin);
      }
    });
    newBackgroundPlugins.forEach((plugin) => {
      if (
        !oldBackgroundPlugins.includes(plugin) &&
        this.shouldConnectAsBackgroundPlugin(plugin)
      ) {
        this.initPlugin(plugin);
      }
    });
    this.emit('plugins-change');
  }

  async deviceSerial(): Promise<string> {
    try {
      const device = await this.device;
      if (!device) {
        console.error('Using "" for deviceId device is not ready');
        return '';
      }
      return device.serial;
    } catch (e) {
      console.error(
        'Using "" for deviceId because client has no matching device',
      );
      return '';
    }
  }

  onMessage(msg: string) {
    if (typeof msg !== 'string') {
      return;
    }

    let rawData;
    try {
      rawData = JSON.parse(msg);
    } catch (err) {
      console.error(`Invalid JSON: ${msg}`, 'clientMessage');
      return;
    }

    const data: {
      id?: number;
      method?: string;
      params?: Params;
      success?: Object;
      error?: ErrorType;
    } = rawData;

    const {id, method} = data;

    if (id == null) {
      const {error} = data;
      if (error != null) {
        console.error(
          `Error received from device ${
            method ? `when calling ${method}` : ''
          }: ${error.message} + \nDevice Stack Trace: ${error.stacktrace}`,
          'deviceError',
        );
        this.device.then((device) => handleError(this.store, device, error));
      } else if (method === 'refreshPlugins') {
        this.refreshPlugins();
      } else if (method === 'execute') {
        invariant(data.params, 'expected params');
        const params: Params = data.params;
        const bytes = msg.length * 2; // string lengths are measured in UTF-16 units (not characters), so 2 bytes per char
        emitBytesReceived(params.api, bytes);

        const persistingPlugin:
          | typeof FlipperPlugin
          | typeof FlipperDevicePlugin
          | undefined =
          this.store.getState().plugins.clientPlugins.get(params.api) ||
          this.store.getState().plugins.devicePlugins.get(params.api);

        if (persistingPlugin && persistingPlugin.persistedStateReducer) {
          const pluginKey = getPluginKey(
            this.id,
            {serial: this.query.device_id},
            params.api,
          );
          flipperRecorderAddEvent(pluginKey, params.method, params.params);
          processMessageLater(this.store, pluginKey, persistingPlugin, params);
        }
        const apiCallbacks = this.broadcastCallbacks.get(params.api);
        if (!apiCallbacks) {
          return;
        }

        const methodCallbacks = apiCallbacks.get(params.method);
        if (methodCallbacks) {
          for (const callback of methodCallbacks) {
            callback(params.params);
          }
        }
      }
      return; // method === 'execute'
    }

    if (this.sdkVersion < 1) {
      const callbacks = this.requestCallbacks.get(id);
      if (!callbacks) {
        return;
      }
      this.requestCallbacks.delete(id);
      this.finishTimingRequestResponse(callbacks.metadata);
      this.onResponse(data, callbacks.resolve, callbacks.reject);
    }
  }

  onResponse(
    data: {
      success?: Object;
      error?: ErrorType;
    },
    resolve: ((a: any) => any) | undefined,
    reject: (error: ErrorType) => any,
  ) {
    if (data.success) {
      resolve && resolve(data.success);
    } else if (data.error) {
      reject(data.error);
      const {error} = data;
      if (error) {
        this.device.then((device) => handleError(this.store, device, error));
      }
    } else {
      // ???
    }
  }

  toJSON(): ClientExport {
    return {id: this.id, query: this.query};
  }

  subscribe(api: string, method: string, callback: (params: Object) => void) {
    let apiCallbacks = this.broadcastCallbacks.get(api);
    if (!apiCallbacks) {
      apiCallbacks = new Map();
      this.broadcastCallbacks.set(api, apiCallbacks);
    }

    let methodCallbacks = apiCallbacks.get(method);
    if (!methodCallbacks) {
      methodCallbacks = new Set();
      apiCallbacks.set(method, methodCallbacks);
    }
    methodCallbacks.add(callback);
  }

  unsubscribe(api: string, method: string, callback: Function) {
    const apiCallbacks = this.broadcastCallbacks.get(api);
    if (!apiCallbacks) {
      return;
    }

    const methodCallbacks = apiCallbacks.get(method);
    if (!methodCallbacks) {
      return;
    }
    methodCallbacks.delete(callback);
  }

  rawCall<T>(method: string, fromPlugin: boolean, params?: Params): Promise<T> {
    return new Promise((resolve, reject) => {
      const id = this.messageIdCounter++;
      const metadata: RequestMetadata = {
        method,
        id,
        params,
      };

      if (this.sdkVersion < 1) {
        this.requestCallbacks.set(id, {reject, resolve, metadata});
      }

      const data = {
        id,
        method,
        params,
      };

      const plugin = params ? params.api : undefined;

      console.debug(data, 'message:call');

      if (this.sdkVersion < 1) {
        this.startTimingRequestResponse({method, id, params});
        if (this.connection) {
          this.connection.fireAndForget({data: JSON.stringify(data)});
        }
        return;
      }

      const mark = this.getPerformanceMark(metadata);
      performance.mark(mark);
      if (!fromPlugin || this.isAcceptingMessagesFromPlugin(plugin)) {
        this.connection &&
          this.connection
            .requestResponse({data: JSON.stringify(data)})
            .subscribe({
              onComplete: (payload) => {
                if (!fromPlugin || this.isAcceptingMessagesFromPlugin(plugin)) {
                  const logEventName = this.getLogEventName(data);
                  this.logger.trackTimeSince(mark, logEventName);
                  emitBytesReceived(
                    plugin || 'unknown',
                    payload.data.length * 2,
                  );
                  const response: {
                    success?: Object;
                    error?: ErrorType;
                  } = JSON.parse(payload.data);

                  this.onResponse(response, resolve, reject);
                }
              },
              // Open fresco then layout and you get errors because responses come back after deinit.
              onError: (e) => {
                if (this.isAcceptingMessagesFromPlugin(plugin)) {
                  reject(e);
                }
              },
            });
      }
    });
  }

  startTimingRequestResponse(data: RequestMetadata) {
    performance.mark(this.getPerformanceMark(data));
  }

  finishTimingRequestResponse(data: RequestMetadata) {
    const mark = this.getPerformanceMark(data);
    const logEventName = this.getLogEventName(data);
    this.logger.trackTimeSince(mark, logEventName);
  }

  isAcceptingMessagesFromPlugin(plugin: string | null | undefined) {
    return this.connection && (!plugin || this.activePlugins.has(plugin));
  }

  getPerformanceMark(data: RequestMetadata): string {
    const {method, id} = data;
    return `request_response_${method}_${id}`;
  }

  getLogEventName(data: RequestMetadata): string {
    const {method, params} = data;
    return params && params.api && params.method
      ? `request_response_${method}_${params.api}_${params.method}`
      : `request_response_${method}`;
  }

  initPlugin(pluginId: string) {
    this.activePlugins.add(pluginId);
    this.rawSend('init', {plugin: pluginId});
  }

  deinitPlugin(pluginId: string) {
    this.activePlugins.delete(pluginId);
    this.rawSend('deinit', {plugin: pluginId});
  }

  rawSend(method: string, params?: Object): void {
    const data = {
      method,
      params,
    };
    console.debug(data, 'message:send');
    if (this.connection) {
      this.connection.fireAndForget({data: JSON.stringify(data)});
    }
  }

  call(
    api: string,
    method: string,
    fromPlugin: boolean,
    params?: Object,
  ): Promise<Object> {
    return reportPluginFailures(
      this.rawCall('execute', fromPlugin, {api, method, params}),
      `Call-${method}`,
      api,
    );
  }

  send(api: string, method: string, params?: Object): void {
    if (!isProduction()) {
      console.warn(
        `${api}:${
          method || ''
        } client.send() is deprecated. Please use call() instead so you can handle errors.`,
      );
    }
    return this.rawSend('execute', {api, method, params});
  }

  async supportsMethod(api: string, method: string): Promise<boolean> {
    if (this.sdkVersion < 2) {
      return Promise.resolve(false);
    }
    const response = await this.rawCall<{
      isSupported: boolean;
    }>('isMethodSupported', true, {
      api,
      method,
    });
    return response.isSupported;
  }
}
