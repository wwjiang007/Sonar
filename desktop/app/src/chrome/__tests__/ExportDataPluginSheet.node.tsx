/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import React from 'react';
import Client from '../../Client';
import {create, act, ReactTestRenderer} from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {default as BaseDevice} from '../../devices/BaseDevice';
import ExportDataPluginSheet from '../ExportDataPluginSheet';
import {FlipperPlugin, FlipperDevicePlugin} from '../../plugin';

function generateClientIdentifier(device: BaseDevice, app: string): string {
  const {os, deviceType, serial} = device;
  const identifier = `${app}#${os}#${deviceType}#${serial}`;
  return identifier;
}

class TestPlugin extends FlipperPlugin<any, any, any> {}
TestPlugin.title = 'TestPlugin';
TestPlugin.id = 'TestPlugin';
TestPlugin.defaultPersistedState = {msg: 'Test plugin'};
class TestDevicePlugin extends FlipperDevicePlugin<any, any, any> {}
TestDevicePlugin.title = 'TestDevicePlugin';
TestDevicePlugin.id = 'TestDevicePlugin';
TestDevicePlugin.defaultPersistedState = {msg: 'TestDevicePlugin'};

function getStore(selectedPlugins: Array<string>) {
  const logger = {
    track: () => {},
    info: () => {},
    warn: () => {},
    error: () => {},
    debug: () => {},
    trackTimeSince: () => {},
  };
  let mockStore = configureStore([])();
  const selectedDevice = new BaseDevice(
    'serial',
    'emulator',
    'TestiPhone',
    'iOS',
  );
  const client = new Client(
    generateClientIdentifier(selectedDevice, 'app'),
    {app: 'app', os: 'iOS', device: 'TestiPhone', device_id: 'serial'},
    null,
    logger,
    // @ts-ignore
    mockStore,
    ['TestPlugin', 'TestDevicePlugin'],
  );

  const pluginStates: {[key: string]: any} = {};
  pluginStates[`${client.id}#TestDevicePlugin`] = {
    msg: 'Test Device plugin',
  };
  pluginStates[`${client.id}#TestPlugin`] = {
    msg: 'Test plugin',
  };
  mockStore = configureStore([])({
    application: {share: {closeOnFinish: false, type: 'link'}},
    plugins: {
      clientPlugins: new Map([['TestPlugin', TestPlugin]]),
      devicePlugins: new Map([['TestDevicePlugin', TestDevicePlugin]]),
      gatekeepedPlugins: [],
      disabledPlugins: [],
      failedPlugins: [],
      selectedPlugins,
    },
    pluginStates,
    pluginMessageQueue: [],
    connections: {selectedApp: client.id, clients: [client]},
  });
  return mockStore;
}

test('SettingsSheet snapshot with nothing enabled', async () => {
  let root: ReactTestRenderer;
  await act(async () => {
    root = create(
      <Provider store={getStore([])}>
        <ExportDataPluginSheet onHide={() => {}} />
      </Provider>,
    );
  });

  expect(root!.toJSON()).toMatchSnapshot();
});

test('SettingsSheet snapshot with one plugin enabled', async () => {
  let root: ReactTestRenderer;
  await act(async () => {
    root = create(
      <Provider store={getStore(['TestPlugin'])}>
        <ExportDataPluginSheet onHide={() => {}} />
      </Provider>,
    );
  });

  expect(root!.toJSON()).toMatchSnapshot();
});
