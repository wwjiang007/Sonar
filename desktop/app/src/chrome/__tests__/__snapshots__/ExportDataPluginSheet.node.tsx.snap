// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsSheet snapshot with nothing enabled 1`] = `
<div
  className="css-1alsdwv"
>
  <div
    className="css-14m1iua"
  >
    <div
      className="css-1k4677w"
    >
      <span
        className="css-15682x"
      >
        Select the plugins for which you want to export the data
      </span>
      <div
        className="css-9dmkpi"
      >
        <div
          className="css-1k4677w"
        >
          <div
            className="css-1obf64m"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="css-1jrm6r3"
            >
              <div
                className="css-a1glw"
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <span
                  className="css-1cegttb"
                >
                  TestDevicePlugin
                </span>
                <div
                  className="css-12zzrdt"
                />
                <input
                  checked={false}
                  className="css-1y77uts"
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
            <div
              className="css-1nj3es"
            />
          </div>
        </div>
        <div
          className="css-1k4677w"
        >
          <div
            className="css-1obf64m"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="css-1jrm6r3"
            >
              <div
                className="css-a1glw"
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <span
                  className="css-1cegttb"
                >
                  TestPlugin
                </span>
                <div
                  className="css-12zzrdt"
                />
                <input
                  checked={false}
                  className="css-1y77uts"
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
            <div
              className="css-1nj3es"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1yqvjo0"
    >
      <div
        className="css-a1glw"
      >
        <div
          className="css-12zzrdt"
        />
        <div
          className="css-12n892b"
        >
          <div
            className="css-1ee9nwd"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
          >
            Close
          </div>
        </div>
        <div
          className="css-1obf64m"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="css-6lxjsf"
            disabled={true}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            type="primary"
          >
            Submit
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SettingsSheet snapshot with one plugin enabled 1`] = `
<div
  className="css-1alsdwv"
>
  <div
    className="css-14m1iua"
  >
    <div
      className="css-1k4677w"
    >
      <span
        className="css-15682x"
      >
        Select the plugins for which you want to export the data
      </span>
      <div
        className="css-9dmkpi"
      >
        <div
          className="css-1k4677w"
        >
          <div
            className="css-1obf64m"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="css-1jrm6r3"
            >
              <div
                className="css-a1glw"
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <span
                  className="css-1cegttb"
                >
                  TestDevicePlugin
                </span>
                <div
                  className="css-12zzrdt"
                />
                <input
                  checked={false}
                  className="css-1y77uts"
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
            <div
              className="css-1nj3es"
            />
          </div>
        </div>
        <div
          className="css-1k4677w"
        >
          <div
            className="css-1obf64m"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="css-1jrm6r3"
            >
              <div
                className="css-a1glw"
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <span
                  className="css-1cegttb"
                >
                  TestPlugin
                </span>
                <div
                  className="css-12zzrdt"
                />
                <input
                  checked={true}
                  className="css-1y77uts"
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
            <div
              className="css-1nj3es"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1yqvjo0"
    >
      <div
        className="css-a1glw"
      >
        <div
          className="css-12zzrdt"
        />
        <div
          className="css-12n892b"
        >
          <div
            className="css-1ee9nwd"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
          >
            Close
          </div>
        </div>
        <div
          className="css-1obf64m"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="css-1ee9nwd"
            disabled={false}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            type="primary"
          >
            Submit
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
